core/src/SYCL/Kokkos_SYCL_Instance.cpp

```diff
@@ -315,7 +328,8 @@ size_t SYCLInternal::USMObjectMem<Kind>::reserve(size_t n) {
         AllocationSpace(*m_q), "Kokkos::Experimental::SYCL::USMObjectMem", n);
     Record::increment(r);
 
-    m_data     = r->data();
+    m_data = r->data();
+    m_staging.reset(new char[n]);
```

Comment:
We only need this if `sycl::usm::alloc::device == Kind`.
```suggestion
    if constexpr (sycl::usm::alloc::device == Kind)
      m_staging.reset(new char[n]);
```